File: 2020/21_AllergenAssessment/shopping.lua
-- ======================================================================
-- Allergen Assessment
--   Advent of Code 2020 Day 21 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         s h o p p i n g . l u a
-- ======================================================================
-- A solver for the Advent of Code 2020 Day 21 puzzle

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local label = require('label')
local set = require('set')

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Shopping = { part2 = false, text = {}, 
  labels = {}, ingredients = {}, allergens = {}, mapping = {} }
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                               Shopping
-- ======================================================================

-- Object for Allergen Assessment

function Shopping:Shopping (o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.labels = {}
  o.ingredients = set:Set()
  o.allergens = set:Set()
  o.mapping = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Shopping:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)

  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do

    -- 2. Add the label to the labels
    table.insert(self.labels, label:Label({part2=self.part2, text=line}))
    end

  -- 2. Get list of ingredients
  for _, lab in ipairs(self.labels) do
    for ing in pairs(lab.ingredients) do
      self.ingredients:add(ing)
    end
  end
  
  -- 3. Get list of allergens
  for _, lab in ipairs(self.labels) do
    for alg in pairs(lab.allergens) do
      self.allergens:add(alg)
    end
  end
  
  -- 4. Map the allergens to (possible) ingredients
  for _, lab in ipairs(self.labels) do
    for alg in pairs(lab.allergens) do
      if nil == self.mapping[alg] then
        self.mapping[alg] = set:Set(lab.ingredients)
      else
        self.mapping[alg] = self.mapping[alg] * lab.ingredients
      end
    end
  end
end

function Shopping:not_safe()
  -- Returns all of the not safe ingredients
  
  -- 1. Start with nothing
  local result = set:Set()
  
  -- 2. Loop for all of the allergens mapped to ingredients
  for _, ings in pairs(self.mapping) do
    
    -- 3. Loop for all of the ingredients in that alergin
    for ing, _ in pairs(ings) do
      
      -- 4. Add to the unsafe ingredients
      result:add(ing)
    end
  end
  
  -- 5. Return the suspected unsafe ingredients
  return result
end

function Shopping:safe()
  -- Returns all of the safe ingredients
  
  -- 1. Start with all the ingredients
  local result = set:Set(self.ingredients)
  
  -- 2. Remove the unsafe ones
  result = result - self:not_safe()
  
  -- 3. Return the safe ingredients
  return result
end

function Shopping:count_ingredient(ingredient)
  -- Return the number of times the ingredient appears
  
  -- 1. Start with nothing
  local result = 0
  
  -- 2. Loop for all of the labels
  for _, lab in ipairs(self.labels) do
    
    -- 3. If the ingredient appears on the label, increment count
    if lab:has_ingredient(ingredient) then
      result = result + 1
    end
  end
  
  -- 4. Return the ingredient count
  return result
end

function Shopping:count_ingredients(ingredients)
  -- Return the number of times the ingredients in the list appear
  
  -- 1. Start with nothing
  local result = 0
  
  -- 2. Loop for all of the ingredients
  for ing in pairs(ingredients) do
    
    -- 3. Add in the number of times this ingredient appears
    result = result + self:count_ingredient(ing)
  end
  
  -- 4. Return the total ingredient count
  return result
end

function Shopping:resolve_allergens()
  -- Returns true if we could resolve the allergens to their ingredients
  
  -- 1. Impose a limit
  for _ = 1, 10 do
    local done = true
    
    -- 2. Once we find an allergen that can be in only one ingredient
    --    we can remove that ingredient from the other allergen lists
    for _, ingredients in pairs(self.mapping) do
      if #ingredients == 1 then
        local ing = ingredients:pop()
        for _, other in pairs(self.mapping) do
          other:remove(ing)
        end
        ingredients:add(ing)
      else
        done = false
      end
      
      -- 3. If all allergens have only one ingredient, we are done
      if done then
        return true
      end
    end
  end
  
  -- 4. All that work and nothing to show for it
  return false
end

function Shopping:sorted_ingredients()
  -- Return the list of ingredients sorted by allergens
  
  -- 1. Start with nothing
  local result = {}
  
  -- 2. Sort the mapping by allergen
  local algs = self.allergens:tolist()
  table.sort(algs)
      
  -- 3. Add the sorted ingredians to the result
  for _, alg in ipairs(algs) do
    table.insert(result, self.mapping[alg]:element())
  end
  
  -- 4. Return the sorted ingredients
  return result
end

function Shopping:part_one(args)
  -- Returns the solution for part one

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:count_ingredients(self:safe())
end


function Shopping:part_two(args)
  -- Returns the solution for part two

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  if self:resolve_allergens() then 
    return table.concat(self:sorted_ingredients(), ',')
  end
  return nil
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Shopping

-- ======================================================================
-- end                     s h o p p i n g . l u a                    end
-- ======================================================================

File: 2020/21_AllergenAssessment/shopping.ts
// ======================================================================
// Allergen Assessment
//   Advent of Code 2020 Day 21 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           s h o p p i n g . t s
//
// A solver for the Advent of Code 2020 Day 21 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------
import { Label } from './label';

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type AllergenToIngredient = Record<string, Set<string>>;

// ======================================================================
//                                                               Shopping
// ======================================================================

export class Shopping {
  // Object for Allergen Assessment
  text: string[];

  part2: boolean;

  labels: Label[];

  ingredients: Set<string>;

  allergens: Set<string>;

  mapping: AllergenToIngredient;

  constructor(text: string[], part2 = false) {
    // Create a Shopping object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.labels = [];
    this.ingredients = new Set();
    this.allergens = new Set();
    this.mapping = {};

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText(this.text);
    }
  }

  processText(text: string[]) {
    // Assign values from text

    // 1. Loop for each line of text
    for (let tindx = 0; tindx < text.length; tindx += 1) {
      // 2. Create a label from the text and add it to the list
      this.labels.push(new Label(text[tindx], this.part2));
    }

    // 2. Get complete list of ingredients
    for (let lindx = 0; lindx < this.labels.length; lindx += 1) {
      const ingredients = this.labels[lindx].getIngredients();
      for (let indx = 0; indx < ingredients.length; indx += 1) {
        this.ingredients.add(ingredients[indx]);
      }
    }

    // 3. Get complete list of alergens
    for (let lindx = 0; lindx < this.labels.length; lindx += 1) {
      const allergens = this.labels[lindx].getAllergens();
      for (let indx = 0; indx < allergens.length; indx += 1) {
        this.allergens.add(allergens[indx]);
      }
    }

    // 4. Create mapping of allergens to ingredients
    for (let lindx = 0; lindx < this.labels.length; lindx += 1) {
      const label = this.labels[lindx];
      const allergens = label.getAllergens();
      for (let aindx = 0; aindx < allergens.length; aindx += 1) {
        const allergen = allergens[aindx];
        if (this.mapping[allergen] === undefined) {
          this.mapping[allergen] = new Set(label.ingredients);
        } else {
          const ingredients = Array.from(this.mapping[allergen]);
          for (let indx = 0; indx < ingredients.length; indx += 1) {
            const ingredient = ingredients[indx];
            if (!label.ingredients.has(ingredient)) {
              this.mapping[allergen].delete(ingredient);
            }
          }
        }
      }
    }
  }

  notSafe(): Set<string> {
    // Return set of ingredients that have allergens

    // 1. Start with nothing
    const result: Set<string> = new Set();

    // 2. Loop for all of the allergens map to ingredients
    const allergins = Array.from(this.allergens);
    for (let aindx = 0; aindx < allergins.length; aindx += 1) {
      const allergin = allergins[aindx];
      // 3. Loop for all of the ingredients possible for that allergen
      const ingredients = Array.from(this.mapping[allergin]);
      for (let indx = 0; indx < ingredients.length; indx += 1) {
        // 4. Add the unsave ingredients to the result
        result.add(ingredients[indx]);
      }
    }
    // 5. Return the set of ingredients that have allergins
    return result;
  }

  safe(): Set<string> {
    // Returns set of ingredients that don't have allergens

    // 1. Start with the set of all ingrdents
    const result = new Set(this.ingredients);

    // 2. Get an array of ingredients with allergens
    const unsafe = Array.from(this.notSafe());

    // 3. Remove the unsage ingredients from the result
    for (let indx = 0; indx < unsafe.length; indx += 1) {
      result.delete(unsafe[indx]);
    }

    // 4. Return the remaining ingredients
    return result;
  }

  countIngredient(ingredient: string): number {
    // Return the number of labels on which the ingredient appears

    // 1. Start with nothing
    let result = 0;

    // 2. Loop for all of the labels
    for (let indx = 0; indx < this.labels.length; indx += 1) {
      const label = this.labels[indx];

      // 3. If this label has the ingredient, increment count
      if (label.ingredients.has(ingredient)) {
        result += 1;
      }
    }

    // 4. Return the number of labels with the ingredient
    return result;
  }

  countIngredients(ingredients: string[]): number {
    // Return the total number of labels on which the ingredients appears

    // 1. Start with nothing
    let result = 0;

    // 2. Loop fpr all of the ingredients given
    for (let indx = 0; indx < ingredients.length; indx += 1) {
      // 3. Add in the number of times this ingrdient appears
      result += this.countIngredient(ingredients[indx]);
    }
    // 4. Return total label count
    return result;
  }

  resolveAllergens(): boolean {
    // Returns true if we could resolve the allergens to thier ingredients

    // 1. Get a list of the known allergens
    const allergens = Array.from(this.allergens);

    // 2. Impose a limit
    for (let limit = 0; limit < 10; limit += 1) {
      let done = true;
      // 3 Once we find an allergin that can be in only one ingredient ...
      for (let aindx = 0; aindx < allergens.length; aindx += 1) {
        const allergen = allergens[aindx];
        const ingredients = Array.from(this.mapping[allergen]);
        if (ingredients.length === 1) {
          const ingredient = ingredients[0];
          // 4. We can remove that ingredient from the other allergen lists
          for (let indx = 0; indx < allergens.length; indx += 1) {
            if (indx !== aindx) {
              this.mapping[allergens[indx]].delete(ingredient);
            }
          }
        } else {
          done = false;
        }
      }
      // 5. if all allergens have only one ingredient, we are done
      if (done) {
        return true;
      }
    }
    // 6. All that work and nothing to show for it
    return false;
  }

  sortedIngredients(): string[] {
    // Returns the list of ingredients sorted by allergens

    // 1. Start with nothing
    const result: string[] = [];

    // 2. Get the allergens in sorted order
    const allergens = Array.from(this.allergens);
    allergens.sort();

    // 3. Loop for the sorted allergens
    for (let indx = 0; indx < allergens.length; indx += 1) {
      // 4. Get the single ingredient
      const ingredient = Array.from(this.mapping[allergens[indx]])[0];
      // 5. Add it to the result
      result.push(ingredient);
    }
    // 6. Return the ingredients sorted by allergens
    return result;
  }

  solution(verbose = false, limit = 0): number|string {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      if (!this.resolveAllergens()) {
        return NaN;
      }
      return this.sortedIngredients().join(',');
    }
    return this.countIngredients(Array.from(this.safe()));
  }

  partOne(verbose = false, limit = 0): number|string {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number|string {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                      s h o p p i n g . t s                     end
// ======================================================================
